public/js/util/source-map.js

```diff
@@ -1,29 +1,112 @@
 "use strict";
 
 const { networkRequest } = require("./networkRequest");
-const { SourceMapConsumer } = require("source-map");
+const invariant = require("invariant");
+const { SourceMapConsumer, SourceNode } = require("source-map");
+const { Source } = require("../types");
+const { Task } = require("./task");
+const { getSource } = require("../selectors");
+const first = require("lodash/first");
 
-let sourceMaps = new Map();
+let sourceMapConsumers = new Map();
+let sourceNodes = new Map();
+
+function hasSourceMap(sourceId) {
+  return sourceMapConsumers.has(sourceId)
+}
+
+function getConsumer(sourceId) {
+  return sourceMapConsumers.get(sourceId);
+}
+
+function getOriginalSources(sourceId) {
+  const consumer = getConsumer(sourceId);
+  return consumer.sources;
+}
+
+function isOriginal(originalSource) {
+  return !!getGeneratedSourceId(originalSource)
+}
+
+function getGeneratedSourceId(originalSource) {
```

Comment:
Really happy with this method. This is the third attempt at mapping generated => original. The first attempt was to add a generatedSourceId to all sources. The second was to create an original source id to generated source id map. This is a simple approach that re-uses the abstractions we already have.